### PR TITLE
Add Nullable module and generics map and bind

### PR DESCRIPTION
**Pull request by mattstermiller**

### CODE REVIEW

File: src/FSharpPlus/Control/Monad.fs [hunk 1]
```diff
@@ -32,6 +32,7 @@ type Bind =
     static member        (>>=) (source             , f: 'T -> _           ) = async.Bind (source, f)                                        : Async<'U>
     static member        (>>=) (source             , k: 'T -> _           ) = Result.bind k source                                          : Result<'U,'E>
     static member        (>>=) (source             , k: 'T -> _           ) = Choice.bind k source                                          : Choice<'U,'E>
+    static member        (>>=) (source             , f: 'T -> _           ) = Nullable.bind   f source                                      : Nullable<'U>
```
Comment (wallymathieu) [inline src/FSharpPlus/Control/Monad.fs:35]: Perhaps remove these overloads in order for the tests to compile?

**Comment by wallymathieu** [comment]
The overloading of map and bind are probably a bit problematic. What do you say about only adding the Nullable module in order to be able to iterate on the nullable overloads later on?

**Comment by mattstermiller** [comment]
Those overloads were the whole reason for the PR. I wanted Nullable support for the generic map and bind operators. I can easily create a Nullable module in my own utility library.

I don't understand the type sorcery going on to even make an educated guess as to how to move forward with this. I can't even get the test project from the tip of master to compile on my machine with the latest dotnet SDK.

**Comment by mattstermiller** [comment]
OK, I just found a comment in another issue stating that debug builds don't compile. It seems like there should be a huge warning somewhere about this. The fact that it doesn't compile is a huge turn-off to a new contributor.

**Comment by wallymathieu** [comment]
What happens if you try to build it using `dotnet build -c Release` ? There is an f# bug that causes the compilation to fail for debug builds.

File: src/FSharpPlus/Control/Functor.fs [hunk 2]
```diff
@@ -81,6 +81,7 @@ type Map =
     static member Map ((x: StringBuilder       , f        ), _mthd: Map) = new StringBuilder (String.map f (string x))
     static member Map ((x: Set<_>              , f        ), _mthd: Map) = Set.map f x
     static member Map ((_: Set2<'T>            , _: 'T->'U), _mthd: Map) = Set2<'U>()
+    static member Map ((x: Nullable<_>         , f        ), _mthd: Map) = Nullable.map f x
```
Comment (wallymathieu) [inline src/FSharpPlus/Control/Functor.fs:84]: The compilation issue is due to the fact that it chooses Nullable over other overloads when resolving the correct overload. If you look further down in the file you'll notice that other overloads have the priority set using `Default1` and `Default2`.

Comment (mattstermiller) [reply src/FSharpPlus/Control/Functor.fs]: Wow, thanks! That works on my machine.

**Comment by wallymathieu** [comment]
The issue around the Fable compilation is due to the fact that nullable does not exist in Fable, why you need to surround nullable module it by an `#if` block.

File: tests/FSharpPlus.Tests/Extensions.fs [hunk 3]
```diff
@@ -267,4 +268,18 @@ module Extensions =
 
   [<Test>]
   let ``Array.lift2 should combine all arrays `` () =
-    areStEqual [|11; 21; 31; 12; 22; 32|] (Array.lift2 (+) [|1;2|] [|10;20;30|])
\ No newline at end of file
+    areStEqual [|11; 21; 31; 12; 22; 32|] (Array.lift2 (+) [|1;2|] [|10;20;30|])
+
+#if !FABLE_COMPILER
```
Comment (wallymathieu) [inline tests/FSharpPlus.Tests/Extensions.fs:273]: This test project is not compiled for Fable unfortunately.

**Comment by gusty** [comment]
What's the status on this?
Is it ready for review?
If the idea is to keep the generic overloads I would add some tests for them.

**Comment by gusty** [comment]
I see there are currently some tests, included in the same test for the modules, that's fine but they seems to be monomorphic (int to int). Can you add polymorphic ones? Like mapping/binding from int to other type?

**Comment by mattstermiller** [comment]
I've added polymorphic functions to the map and bind tests and added a test for `iter`.

If you would prefer me to move the generic function tests, please let me know where they should go.

Thank you for your interest in this request :)

**Comment by gusty** [comment]
That makes sense. I'll review it and most likely I will accept it.
The only part that I'm a bit unsure is whether to include the generic versions in v1.1 because we already did 4 RCs and I don't want to risk adding something new just for the release.
But will see what can I do.

Thanks for contributing back !

**Comment by mattstermiller** [comment]
@gusty I have another commit that adds more functions to this Nullable module for #191. Would you like me to do a separate PR for that, or just push another commit to this one?

**Comment by gusty** [comment]
Great, do it here.
Later we can always either rename this PR, or separate in a second one.

**Pull request merged**